### DISCUSSION TITLE
gtsummary stopped working with targets when Rmd not placed in main folder

### DISCUSSION THREAD

**Discussion by gorkang** [ropensci/targets#920] [Help]
## Prework

* [x] Read and agree to the [code of conduct](https://ropensci.org/code-of-conduct/) and [contributing guidelines](https://github.com/ropensci/targets/blob/main/CONTRIBUTING.md).
* [ ] Confirm that your issue is a genuine bug in the `targets` package itself and not a user error, known limitation, or issue from another package that `targets` depends on. For example, if you get errors running `tar_make_clustermq()`, try isolating the problem in a reproducible example that runs `clustermq` and not `targets`. And for miscellaneous troubleshooting, please post to [discussions](https://github.com/ropensci/targets/discussions) instead of [issues](https://github.com/ropensci/targets/issues).
* [x] If there is [already a relevant issue](https://github.com/ropensci/targets/issues), whether open or closed, comment on the existing thread instead of posting a new issue.
* [x] Using [`targets::tar_reprex()`](https://docs.ropensci.org/targets/reference/tar_reprex.html), [`reprex::reprex()`](https://reprex.tidyverse.org/reference/reprex.html), or similar, post a [minimal reproducible example](https://www.tidyverse.org/help/) like [this one](https://github.com/ropensci/targets/issues/256#issuecomment-754229683) so the maintainer can troubleshoot the problems you identify. A reproducible example is:
    * [x] **Runnable**: post enough R code and data so any onlooker can create the error on their own computer.
    * [x] **Minimal**: reduce runtime wherever possible and remove complicated details that are irrelevant to the issue at hand.
    * [x] **Readable**: format your code according to the [tidyverse style guide](https://style.tidyverse.org/).

## Description

Not sure if I should post this bug report here or in `gtsummary`, or `tarchetypes`. Sorry if this is not the right place. I tried using previous versions of targets, tarchetypes and gtsummary to locate the source of the problem, but the error persisted.

In brief, if I use  `gtsummary::tbl_summary()` in a .Rmd in a targets pipeline, and the .Rmd is not in the main folder, the pipeline fails 

## Reproducible example

This is a project with a minimal reproducible example: [targets_gtsummary.zip](https://github.com/ddsjoberg/gtsummary/files/9470163/targets_gtsummary.zip)

If you run it, it will fail with:

```
Quitting from lines 23-24 (report.Rmd) 
✖ error target report_analysis
• end pipeline: 1.295 seconds
Error : cannot open the connection
✖ Problem with the pipeline.
Error:
! problem with the pipeline.
```

### How to avoid the error

- If you comment line 23 in Rmd/report.Rmd, it will work.

- If you move Rmd/report.Rmd to the main project folder and adjust the tar_render() call, it will work.

- If you create a copy of Rmd/report.Rmd in the rmain project folder, it will work! (will render the Rmd/report.Rmd file though)

## Diagnostic information

R version 4.2.1 (2022-06-23)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.1 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=es_CL.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=es_CL.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=es_CL.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=es_CL.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] pillar_1.8.1        compiler_4.2.1      targets_0.13.1      tools_4.2.1         digest_0.6.29      
 [6] gtable_0.3.0        evaluate_0.16       lifecycle_1.0.1     tibble_3.1.8        pkgconfig_2.0.3    
[11] rlang_1.0.4         igraph_1.3.4        gtsummary_1.6.1     cli_3.3.0.9000      DBI_1.1.3          
[16] rstudioapi_0.13     yaml_2.3.5          xfun_0.32           fastmap_1.1.0       stringr_1.4.1      
[21] withr_2.5.0         dplyr_1.0.9         knitr_1.40          generics_0.1.3      vctrs_0.4.1        
[26] grid_4.2.1          tidyselect_1.1.2    glue_1.6.2          data.table_1.14.2   R6_2.5.1           
[31] processx_3.7.0      fansi_1.0.3         base64url_1.4       rmarkdown_2.16      tidyr_1.2.0        
[36] ggplot2_3.3.6       callr_3.7.2         purrr_0.3.4         magrittr_2.0.3      scales_1.2.1       
[41] backports_1.4.1     codetools_0.2-18    ps_1.7.1            htmltools_0.5.3     broom.helpers_1.8.0
[46] gt_0.7.0            assertthat_0.2.1    colorspace_2.0-3    utf8_1.2.2          stringi_1.7.8      
[51] munsell_0.5.0

**Comment by wlandau** [comment]
Hmm... I ran [targets_gtsummary.zip](https://github.com/ddsjoberg/gtsummary/files/9470163/targets_gtsummary.zip) on R 4.2.1 (Mac OS 12.5.1) and R 4.1.2 (RHEL 7.9) and both times completed without error with line 23 uncommented in the report. Anything more specific about the error on your end? Development `targets` prints out error messages a bit better, for what it's worth.

**Reply by wlandau** [reply]
Or you could try https://books.ropensci.org/targets/debugging.html#the-debug-option, maybe a `tar_target(report, rmarkdown::render("Rmd/report.Rmd"))` combined with `tar_option_set(debug = "report")` and `tar_make(callr_function = NULL)`.

**Comment by gorkang** [comment]
Thanks @wlandau ,

I am trying different options... and if I am not mistaken, some signs seem to point to tarchetypes.

## DEBUG
With development targets, using `tar_option_set(debug = "report")` and `tar_make(callr_function = NULL)` stops with DF in the environment. If I run `gtsummary::tbl_summary(DF)` in the console, all works fine. If I continue, the same error: Quitting from lines 23-24 (report.Rmd) 

## Older versions
As this exact pipeline used to work a few months ago, I tried running it with older versions of RStudio and the packages I thought could be causing issues, but it didn't work either. I get the same error:

```
RStudio 2022.02.3+492
devtools::install_github("rstudio/rmarkdown@v2.12")
devtools::install_github("yihui/knitr@v1.39")
devtools::install_github("ropensci/tarchetypes@0.5.0")
devtools::install_github("ropensci/targets@0.10.0")
devtools::install_github("ddsjoberg/gtsummary@v1.5.2")
```

## rmarkdown::render

Again with development targets, after adding `params = list(DF = DF)` so the render node would be connected to the DF node, using `targets::tar_target(report, rmarkdown::render("Rmd/report.Rmd", params = list(DF = DF)))`, targets::tar_meta() shows the following error: 
- `31mmissing files _targetsmetameta39m`

But, when replace `targets::tar_load(DF)` for `DF = readr::read_rds("../_targets/objects/DF")`, **it works!** It seems this could be an issue with `tarchetypes::tar_render()`?

**Reply by gorkang** [reply]
By the way, tried to run the project in a Windows 10 VM, and it gave the same error.

**Reply by wlandau** [reply]
Try `targets::tar_load(DF, store = file.path(rprojroot::find_rstudio_root_file(), "_targets"))`.

**Reply by wlandau** [reply]
`tar_make()` runs at the project root and creates the `_targets/` data store at the project root. If you nest the report in a different directory, it has to know where to find `_targets/`. This is not a bug, it's a consequence of a deliberate guardrail: `targets` does not search parent directories for `_targets/` because that would cause confusion if a parent directory has a data store left over from another project.

**Comment by gorkang** [comment]
Replacing `targets::tar_load(DF)` for `targets::tar_load(DF, store = file.path(rprojroot::find_rstudio_root_file(), "_targets"))` in Rmd/report.Rmd does not make any difference. 

Keep in mind that DF is accessible without issues in report.Rmd. It is specifically the line with `gtsummary::tbl_summary(DF)` the one that fails, and only if using `tarchetypes::tar_render()`. It works with `rmarkdown::render()`.

**Reply by wlandau** [reply]
From your hint about Windows, I tracked it down to https://github.com/rstudio/gt/blob/a585bd6a916cca767c7a2e3e72eda8f91c2f67f5/R/print.R#L27 (using `tar_option_set(workspace_on_error = TRUE)`, then `tar_traceback(report_analysis)` to get the traceback, then `debug(gt:::knitr_is_word_output)` and `tar_make(callr_function = NULL)`). Looks like `gt` expects the report path to be `report.Rmd` instead of `Rmd/report.Rmd`. This can be fixed if  https://github.com/rstudio/gt/blob/a585bd6a916cca767c7a2e3e72eda8f91c2f67f5/R/print.R#L27 uses `knitr::current_input(dir = TRUE)` instead of `knitr::current_input()`.

**Reply by wlandau** [reply]
I can reproduce the same error without `targets` using `rmarkdown::render("Rmd/report.Rmd", knit_root_dir = getwd())` and the following report:

    ---
    title: "report"
    date: "`r Sys.Date()`"
    output: html_document
    editor_options: 
      chunk_output_type: console
    ---
    
    ```{r setup, include=FALSE}
    knitr::opts_chunk$set(echo = TRUE)
    ```
    
    ## gt table
    
    ```{r table}
    gt::gt(data.frame(x = sample.int(100), y = sample.int(100)))
    ```

`tarchetypes::tar_render()` needs to set `knitr_root_dir = getwd()` to ensure the working directory does not move around when the pipeline runs.

**Comment by gorkang** [comment]
You are completely right! Using gt v0.6 solves the issue:

devtools::install_github("rstudio/gt@v0.6.0")
